commands/common_test.go

```diff
@@ -164,3 +165,412 @@ func (*CMDTest) TestMockContext(c *gc.C) {
 	c.Assert(ctx.GlobalString(tenetCfgFlg.long), gc.Equals, "custom/cfg/path")
 	c.Assert(ctx.Command.Name, gc.Equals, "add")
 }
+
+var complexConfig = &config{
+	Cascade:  true,
+	Include:  "*",
+	Template: "template.md",
+	TenetGroups: []TenetGroup{
+		{
+			Name: "default",
+			Tenets: []TenetConfig{
+				{
+					Name:    "lingoreviews/tenetseed:latest",
+					Options: map[string]interface{}{"opt1": true, "opt2": "an option"},
+				},
+			},
+		},
+		{
+			Name:     "Group1",
+			Template: "g1template.md",
+			Tenets: []TenetConfig{
+				{
+					Name:    "lingo-reviews/license",
+					Options: map[string]interface{}{"header": "// Copyright 2016\n"},
+				},
+				{
+					// Note this is in Group1 and Group2
+					Name:    "lingo-reviews/unused_function_args",
+					Options: map[string]interface{}{},
+				},
+			},
+		},
+		{
+			Name:     "Group2",
+			Template: "g2template.md",
+			Tenets: []TenetConfig{
+				{
+					Name:    "lingoreviews/space_after_forward_slash",
+					Options: map[string]interface{}{},
+				},
+				{
+					// Note this is in Group1 and Group2
+					Name:    "lingo-reviews/unused_function_args",
+					Options: map[string]interface{}{},
+				},
+			},
+		},
+	},
+}
+
+func (*CMDTest) TestAllTenets(c *gc.C) {
+	// AllTenets Should provide a list of all tenets flattened from groups and without duplicates
+	output := complexConfig.AllTenets()
+
+	expected := []TenetConfig{
+		{
+			Name:    "lingoreviews/tenetseed:latest",
+			Options: map[string]interface{}{"opt1": true, "opt2": "an option"},
+		},
+		{
+			Name:    "lingo-reviews/license",
+			Options: map[string]interface{}{"header": "// Copyright 2016\n"},
+		},
+		{
+			Name:    "lingo-reviews/unused_function_args",
+			Options: map[string]interface{}{},
+		},
+		{
+			Name:    "lingoreviews/space_after_forward_slash",
+			Options: map[string]interface{}{},
+		},
+	}
+
+	c.Assert(output, gc.DeepEquals, expected)
+}
+
+func (*CMDTest) TestHasTenetGroup(c *gc.C) {
+	// HasTenetGroup should return true only for existing group names, case sensitive
+	c.Assert(complexConfig.HasTenetGroup("default"), gc.Equals, true)
+	c.Assert(complexConfig.HasTenetGroup("Group1"), gc.Equals, true)
+	c.Assert(complexConfig.HasTenetGroup("Group2"), gc.Equals, true)
+	c.Assert(complexConfig.HasTenetGroup("invalid"), gc.Equals, false)
+	c.Assert(complexConfig.HasTenetGroup("group1"), gc.Equals, false)
+}
+
+func (*CMDTest) TestAddTenetGroup(c *gc.C) {
+	// AddTenetGroup should add a new empty group to the config if that group doesn't already exist
+	testConfig := deepcopy.Iface(complexConfig).(*config)
```

Comment:
I investigated this a bit: I think there's no way to avoid a deep copy (either by package or manually). For the Test*Group tests it's ok but eg. TestAddTenet mutates the config deeply at references which remain shared. I couldn't even resolve it by splitting out to seperate test functions as the one `complexConfig` is shared for the entire test run, and the function call order is undefined.